File: prototype/lib/moof/token.rb
module Moof
  Token = Data.define(:type, :lexeme, :literal, :line, :column) do
    def to_s
      "#{type}(#{lexeme.inspect})"
    end
  end

  # Token type constants
  module TokenType
    LPAREN    = :LPAREN      # (
    RPAREN    = :RPAREN      # )
    LBRACKET  = :LBRACKET    # [
    RBRACKET  = :RBRACKET    # ]
    LBRACE    = :LBRACE      # {
    RBRACE    = :RBRACE      # }

    INTEGER   = :INTEGER     # 42, 0xFF
    FLOAT     = :FLOAT       # 3.14, 1e10
    STRING    = :STRING      # "hello"
    TRUE      = :TRUE        # true
    FALSE     = :FALSE       # false
    NIL       = :NIL         # nil

    IDENTIFIER = :IDENTIFIER # foo, +, my-var
    COLON_ID   = :COLON_ID   # insertValue: (keyword in message sends)

    QUOTE     = :QUOTE       # '
    DOT       = :DOT         # . (rest parameter separator)

    # New tokens
    PIPE          = :PIPE          # |
    BACKTICK      = :BACKTICK      # `
    COMMA         = :COMMA         # ,
    COMMA_AT      = :COMMA_AT      # ,@
    AMPERSAND     = :AMPERSAND     # &
    INTERP_STRING = :INTERP_STRING # $"..."

    EOF       = :EOF
  end
end

File: src/token.rs
#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Delimiters
    LParen, RParen,
    LBracket, RBracket,
    LBrace, RBrace,

    // Literals
    Integer(i64),
    Float(f64),
    Str(String),
    True, False, Nil,

    // Identifiers and keywords
    Identifier(String),
    ColonId(String),    // keyword selector: "insertValue:"

    // Special
    Quote,              // '
    Dot,                // .
    Pipe,               // |
    Backtick,           // `
    Comma,              // ,
    CommaAt,            // ,@
    Ampersand,          // &
    InterpString(Vec<InterpSegment>), // $"..."

    Eof,
}

#[derive(Debug, Clone, PartialEq)]
pub enum InterpSegment {
    Str(String),
    Expr(String),       // source code to be re-parsed
}

#[derive(Debug, Clone)]
pub struct Token {
    pub ty: TokenType,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(ty: TokenType, line: usize, column: usize) -> Self {
        Token { ty, line, column }
    }
}
